### PR TITLE
feat: VM interpreter refactoring and BCHN script test integration

### PR DESCRIPTION
**Pull request by fpelliccioni**
## Summary

Refactor the script interpreter and integrate BCHN's script test battery as a foundation for implementing pending CHIPs (64-bit integers, Native Introspection, BigInt, VM Limits).

### Interpreter refactoring
- Add `op_result` struct pairing generic error categories with failing opcode
- Replace ~90 op-specific error codes with 10 generic categories
- Convert all interpreter `op_xxx` functions to return `{category, opcode}`
- Add `is_signature_result()` to centralize fatal vs non-fatal sig errors
- Fix BCH multisig Satoshi bug handling (always discard dummy element)
- Fix P2SH push-only check to return `sig_pushonly` (matching BCHN)
- Deduplicate `script::verify` by delegating to free `verify()` function

### New feature implementations
- Add `verify_schnorr()` using `secp256k1_schnorr_verify`
- Implement Schnorr signature path in checksig (64-byte endorsements)
- Implement full `checkdatasig`/`checkdatasigverify` with DER and Schnorr support
- Add NULLFAIL enforcement in checkdatasig (`bch_daa_cw144`)
- Add `sig_badlength` rejection for 65-byte endorsements in checkmultisig
- Add MINIMALNUM validation in CLTV/CSV opcodes (`bch_mersenne`)
- Fix CLTV MINIMALNUM check ordering (before `is_final`)
- Fix checkmultisig error propagation for `pop_int32` failures

### Native introspection opcodes
- Implement all 14 introspection opcodes (OP_INPUTINDEX through OP_UTXOBYTECODE)
- Add CashTokens token prefix parsing for `output_token_*` opcodes
- Add `script_execution_context` for pre-computed transaction introspection data
- Propagate context across program constructors

### Signature validation (BCHN-compatible)
- Add `is_valid_signature_encoding()`: pure DER format check matching BCHN
- Add `check_low_s()`: secp256k1-based LOW_S validation
- Add `check_transaction_signature_encoding()`: DER + LOW_S + hashtype + forkid
- Add `check_data_signature_encoding()`: for OP_CHECKDATASIG (no sighash byte)
- Schnorr 65-byte endorsements skip DER/LOW_S checks
- SIGPUSHONLY check before script evaluation in `verify()`
- MINIMALIF enforcement in `op_if`/`op_notif`

### rule_fork refactoring
- Each BCH consensus feature gets its own bit in `forks_t` (`uint64_t`)
- Upgrade groupings as `constexpr` OR combinations
- Add `forks_t` type alias and `are_enabled()` variadic template helper
- Consolidated `missing_forkid` into `must_use_forkid` (BCHN alias)
- Regenerated C-API error codes

### Program and verify
- `program` constructor accepts `value` param for SIGHASH_FORKID amount
- `verify()` passes value through to program for correct sighash computation
- Propagate metrics across program constructors (input → prevout → P2SH)

### INPUT_SIGCHECKS (Fermat upgrade)
- Implement per-input sigchecks density limit: `scriptSig.size() >= sigChecks * 43 - 60`
- ECDSA CHECKMULTISIG counts `nKeysCount` (not 1) when any sig is non-null
- Schnorr CHECKMULTISIG counts 1 per verified signature
- Add `sigchecks_input_density_factor`/`offset` constants

### BigInt (big_number) implementation
- Add `big_number` class wrapping GMP (`mpz_t`) for arbitrary-precision integers
- Support construction from decimal strings, hex strings, and script number serialization
- Implement arithmetic (+, -, *, /, %), bitwise (&, |, ^, <<, >>), and comparison operators
- Add `math_modulo` (floored), `pow`, `pow_mod` operations
- Add `from_hex()` for OpenSSL-format hex parsing
- 23K+ auto-generated tests from BCHN test vectors (`bigint_test_vectors.json`)
- 2.5K+ OpenSSL-format tests (sum, mul/square/quotient, exp, mod, shift)
- Test generators use static arrays + for loops to minimize linker symbols

### VM Limits tests
- Auto-generate 165 VM limits tests from BCHN (128 success + 15 fail + 24 checker)
- Auto-resolve CHECKSIG tests with real transaction context and UTXO cache
- Support CashTokens token prefix parsing in test scriptPubKey

### VM Limits metrics fixes
- Move `sig_checks`/`hash_iterations` tallying into `op_check_sig_common` and `op_check_data_common` (before NULLFAIL check), matching BCHN behavior
- Fix CHECKDATASIG: tally `sig_checks(1)` and `hash_iterations` for any non-empty signature, regardless of verification success/failure
- Fix CHECKSIG: same tallying fix for non-empty failed signatures
- Callers (`op_check_sig`, `op_check_data_sig`, etc.) now only tally `add_op_cost`
- Fix `version_` not propagated in program copy/move constructors (non-BCH)
- Replace `DummySigChecker` with `StandaloneSigChecker` in `eval_script_with_metrics` so CHECKDATASIG does real secp256k1 verification in cross-check tests
- Add CHECKDATASIG failed-verification cross-check test (native vs BCHN)
- Add CHECKDATASIG metrics test for non-empty failed sig (`sig_checks=1`) and empty sig (`sig_checks=0`)

### Additional fixes
- Fix `op_if`/`op_notif`: return `unbalanced_conditional` on empty stack
- Fix `op_check_sequence_verify`: return `insufficient_main_stack` on empty stack
- Fix `op_check_sig_verify`: use `is_signature_result` pattern
- Fix `parse_endorsement`: add `const&` overload to prevent move in ECDSA multisig retry loop

### Test infrastructure
- Integrate BCHN `script_tests.json` with `generate_script_tests.py`
- Generate individual Catch2 `TEST_CASE`s split across 18 files (100/file)
- Test names include expected result: `[Should Pass]` / `[Should Fail: ERROR]`
- Generator validates overrides don't change OK↔FAIL semantics
- Generator handles `[[amount], ...]` format from BCHN for SIGHASH_FORKID tests
- Add override annotation comments to generated test files
- `bchn_script_test` struct supports `amount` field for FORKID tests
- Update `SIG_DER` mapping to `invalid_signature_lax_encoding`
- **All 1641 BCHN script tests passing** (145 overrides, 100 skipped)

## Test plan
- [x] Chunks 0–17 pass (1641 test cases, 4923 assertions)
- [x] DER validation, LOW_S, hashtype, SIGPUSHONLY, SIGHASH_FORKID amount
- [x] Schnorr multisig, SigChecks, INPUT_SIGCHECKS density limit
- [x] IF/NOTIF empty stack, CSV empty stack, CHECKSIGVERIFY
- [x] 165 VM limits tests passing (128 success + 15 fail + 24 checker)
- [x] 23K+ BigInt tests passing (bigint_test_vectors + OpenSSL format)
- [x] Native introspection opcodes (14 opcodes, CashTokens token parsing)
- [x] CHECKDATASIG metrics cross-check (native vs BCHN real verification)
- [x] Re-enable C-API compilation
- [x] Full IBD on mainnet without checkpoints (full validation of all scripts since genesis) — with WITH_CONSENSUS (BCHN VM)
- [x] Full IBD on mainnet without checkpoints (full validation of all scripts since genesis) — without WITH_CONSENSUS (Knuth native VM)
- [x] Full IBD on mainnet with checkpoints (light validation) — with WITH_CONSENSUS (BCHN VM)
- [x] Full IBD on mainnet with checkpoints (light validation) — without WITH_CONSENSUS (Knuth native VM)

---

> [!NOTE]
> **Medium Risk**
> Touches consensus-critical validation paths by changing how script/consensus flags are represented and passed into `block`/`transaction` acceptance and script verification, which could affect rule activation or limits. Risk is mitigated by adding a dedicated VM-limits test suite and optional cross-checking against `WITH_CONSENSUS`.
> 
> **Overview**
> Updates blockchain validation to use `domain::script_flags_t` everywhere instead of legacy fork bitmasks, renaming APIs like `enabled_forks()` to `enabled_flags()` and propagating the new flags through mempool/txpool entries, chain-state population, block/tx acceptance, and `validate_input::verify_script` (including consensus flag conversion).
> 
> Adds a new Catch2-based **VM-Limits** test executable (`kth_blockchain_test`) that runs native VM metric vectors and, when built with `WITH_CONSENSUS`, cross-checks results against BCHN’s consensus interpreter; also introduces build helpers (`scripts/build-native.sh`, `scripts/rebuild-native.sh`) and makes `gmp` a non-optional Conan dependency for native BigInt support.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 37091a457a9c737baca59db1ade0417c4e7bf77b. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/domain/include/kth/domain/impl/machine/operation.ipp [hunk 1]
```diff
@@ -316,49 +315,35 @@ bool operation::is_reserved(opcode code) {
...
+        // case opcode::div:
+        // case opcode::mod:
+        // case opcode::and_:
+        // case opcode::or_:
+        // case opcode::xor_:
+        //     return ! is_enabled(active_forks, rule_fork::bch_pythagoras);
+
+        // div/mod/and/or/xor were reactivated in the 2018-May (pythagoras) upgrade.
+        // Previously gated by bch_reactivated_opcodes, but BCHN has no flag for these
+        // (they fall through to default:false in IsOpcodeDisabled), so we match that behavior.
```
[the remark this answers is not in the document]
Comment (fpelliccioni) [reply src/domain/include/kth/domain/impl/machine/operation.ipp:341]: This is a **false positive**. The current code is correct and intentionally matches BCHN behavior.

In BCHN's `IsOpcodeDisabled()` (`interpreter.cpp:130-145`), these opcodes are **not listed** — they fall through to `default: false`, meaning they are always enabled. They were reactivated in the May 2018 (Pythagoras) upgrade, which is long past on all networks.

Knuth does the same: the cases are commented out with an explicit note explaining this design decision.

File: src/infrastructure/src/math/elliptic_curve.cpp [hunk 2]
```diff
@@ -320,6 +342,22 @@ bool sign_schnorr(ec_signature& out, ec_secret const& secret, hash_digest const&
...
+    secp256k1_pubkey pubkey;
+    auto const context = verification.context();
+    auto const size = point.size();
+
+    if (secp256k1_ec_pubkey_parse(context, &pubkey, point.data(), size) != 1) {
+        return false;
+    }
+
+    return secp256k1_schnorr_verify(context, signature.data(), hash.data(), &pubkey) == 1;
+}
```
[the remark this answers is not in the document]
Comment (fpelliccioni) [reply src/infrastructure/src/math/elliptic_curve.cpp:358]: This is a **false positive**. `secp256k1_ec_pubkey_parse` only triggers the illegal-argument callback when `input` is NULL (`ARG_CHECK(input != NULL)`). Malformed pubkey data simply makes it return 0 — no abort.

Adding `is_public_key()` as a guard would be **incorrect**: BCHN's `CheckPubKeyEncoding` only validates pubkey format when `STRICTENC` is enabled. Without STRICTENC, malformed pubkeys reach `secp256k1_ec_pubkey_parse` directly, which returns 0 → signature fails → `false` pushed to stack → script succeeds. This is consensus behavior.

Knuth already mirrors BCHN exactly: `check_pubkey_encoding()` validates `is_compressed_or_uncompressed_pubkey()` only when `bch_strictenc` is enabled, matching BCHN's `CheckPubKeyEncoding` with `SCRIPT_VERIFY_STRICTENC`.

File: src/c-api/include/kth/capi/error.h [hunk 3]
```diff
@@ -2,6 +2,9 @@
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
+// This file is auto-generated by generate_capi_error.py
+// Do not edit manually.
```
[the remark this answers is not in the document]
Comment (fpelliccioni) [reply src/c-api/include/kth/capi/error.h]: C-API compilation has been re-enabled and the project compiles successfully with all these changes, including the regenerated error enum and the new rule_fork header (now auto-generated via generate_capi_rule_fork.py). The full uint32_t → uint64_t migration across the C-API is also done and validated.

File: src/blockchain/src/settings.cpp [hunk 4]
```diff
@@ -89,21 +89,21 @@ uint32_t settings::enabled_forks() const {
...
+    forks |= (bch_euclid      ? rule_fork::bch_euclid_upgrade : 0);
+    forks |= (bch_pisano      ? rule_fork::bch_pisano_upgrade : 0);
+    forks |= (bch_mersenne    ? rule_fork::bch_mersenne_upgrade : 0);
+    forks |= (bch_fermat      ? rule_fork::bch_fermat_upgrade : 0);
+    forks |= (bch_euler       ? rule_fork::bch_euler_upgrade : 0);
+    forks |= (bch_gauss       ? rule_fork::bch_gauss_upgrade : 0);
+    forks |= (bch_descartes   ? rule_fork::bch_descartes_upgrade : 0);
+    forks |= (bch_lobachevski ? rule_fork::bch_lobachevski_upgrade : 0);
+    forks |= (bch_galois      ? rule_fork::bch_galois_upgrade : 0);
+    // forks |= (bch_leibniz     ? rule_fork::bch_leibniz_upgrade : 0);
```
[the remark this answers is not in the document]
Comment (fpelliccioni) [reply src/blockchain/src/settings.cpp]: This is intentional. Leibniz is the May 2026 upgrade — it's not active yet. The setting is commented out with a placeholder following the same pattern used for all future upgrades (see bch_cantor below it). The chain_state activation code handles the time-based activation independently. This will be uncommented when the upgrade features are implemented.

File: src/c-api/include/kth/capi/chain/transaction.h [hunk 5]
```diff
@@ -105,13 +105,13 @@ kth_input_list_t kth_chain_transaction_inputs(kth_transaction_t transaction);
...
+// code verify(transaction const& tx, uint32_t input_index, uint64_t forks, script const& input_script, script const& prevout_script, uint64_t /*value*/);
 KTH_EXPORT
-kth_error_code_t kth_chain_transaction_verify(kth_transaction_t transaction, uint32_t input_index, uint32_t forks, kth_script_t input_script, kth_script_t prevout_script, uint64_t value);
+kth_error_code_t kth_chain_transaction_verify(kth_transaction_t transaction, uint32_t input_index, uint64_t forks, kth_script_t input_script, kth_script_t prevout_script, uint64_t value);
 
-// code verify(transaction const& tx, uint32_t input, uint32_t forks);
+// code verify(transaction const& tx, uint32_t input, uint64_t forks);
 KTH_EXPORT
-kth_error_code_t kth_chain_transaction_verify_transaction(kth_transaction_t transaction, uint32_t input, uint32_t forks);
+kth_error_code_t kth_chain_transaction_verify_transaction(kth_transaction_t transaction, uint32_t input, uint64_t forks);
```
[the remark this answers is not in the document]
Comment (fpelliccioni) [reply src/c-api/include/kth/capi/chain/transaction.h:114]: Not an issue.

The C-API is versioned with each release and there are no prebuilt consumers that depend on the previous ABI. 

The uint32_t → uint64_t change in forks is intentional and required to support the expanded fork flags.

File: src/blockchain/include/kth/blockchain/populate/populate_base.hpp [hunk 6]
```diff
@@ -23,7 +23,7 @@ class KB_API populate_base {
     populate_base(dispatcher& dispatch, fast_chain const& chain);
 
     void populate_duplicate(size_t maximum_height, const domain::chain::transaction& tx, bool require_confirmed) const;
-    void populate_pooled(domain::chain::transaction const& tx, uint32_t forks) const;
+    void populate_pooled(domain::chain::transaction const& tx, domain::forks_t forks) const;
```
[the remark this answers is not in the document]
Comment (fpelliccioni) [reply src/blockchain/include/kth/blockchain/populate/populate_base.hpp]: Pre-existing bug — the comparison `height == forks` was already present on master with `uint32_t` forks. The parameter was always a fork bitmask, never a height. This PR only widened the type from `uint32_t` to `uint64_t`. Tracked as #190.

File: src/blockchain/src/settings.cpp [hunk 7]
```diff
@@ -89,21 +91,21 @@ uint32_t settings::enabled_forks() const {
...
+    forks |= (bch_pythagoras  ? to_forks(upgrade::bch_pythagoras)  : 0);
+    forks |= (bch_euclid      ? to_forks(upgrade::bch_euclid)      : 0);
+    forks |= (bch_pisano      ? to_forks(upgrade::bch_pisano)      : 0);
+    forks |= (bch_mersenne    ? to_forks(upgrade::bch_mersenne)    : 0);
+    forks |= (bch_fermat      ? to_forks(upgrade::bch_fermat)      : 0);
+    forks |= (bch_euler       ? to_forks(upgrade::bch_euler)       : 0);
+    forks |= (bch_gauss       ? to_forks(upgrade::bch_gauss)       : 0);
+    forks |= (bch_descartes   ? to_forks(upgrade::bch_descartes)   : 0);
+    forks |= (bch_lobachevski ? to_forks(upgrade::bch_lobachevski) : 0);
+    forks |= (bch_galois      ? to_forks(upgrade::bch_galois)      : 0);
```
[the remark this answers is not in the document]
Comment (fpelliccioni) [reply src/blockchain/src/settings.cpp]: This is the intended behavior. BCH network upgrades are cumulative — each upgrade includes all features from prior upgrades. It's not valid to have `bch_galois = true` with `bch_uahf = false`, as Galois inherently requires all previous upgrades to be active. The `to_flags(upgrade::bch_X)` API reflects this by design. The individual bool settings control the maximum upgrade level, not independent feature toggles.

[the remark this answers is not in the document]
Comment (fpelliccioni) [reply src/blockchain/src/settings.cpp]: This is the intended behavior. BCH network upgrades are cumulative — each upgrade includes all features from prior upgrades. It's not valid to have `bch_galois = true` with `bch_uahf = false`, as Galois inherently requires all previous upgrades to be active. The `to_flags(upgrade::bch_X)` API reflects this by design. The individual bool settings control the maximum upgrade level, not independent feature toggles.

File: src/blockchain/src/populate/populate_base.cpp [hunk 8]
```diff
@@ -42,7 +42,7 @@ void populate_base::populate_pooled(const domain::chain::transaction& tx, uint32
         && (position == position_max)) {
 
         tx.validation.pooled = true;
-        tx.validation.current = (height == forks);
+        tx.validation.current = (height == flags);
```
[the remark this answers is not in the document]
Comment (fpelliccioni) [reply src/blockchain/src/populate/populate_base.cpp]: Fixed in PR #192 (merged into master). This branch now includes the fix after rebase — `populate_pooled` correctly compares `stored_height == height` instead of `height == forks`.

[the remark this answers is not in the document]
Comment (fpelliccioni) [reply src/blockchain/src/populate/populate_base.cpp]: Fixed in PR #192 (merged into master). This branch now includes the fix after rebase — `populate_pooled` correctly compares `stored_height == height` instead of `height == forks`.

[the remark this answers is not in the document]
Comment (fpelliccioni) [reply src/blockchain/src/populate/populate_base.cpp]: Fixed in PR #192 (merged into master). This branch now includes the fix after rebase — `populate_pooled` correctly compares `stored_height == height` instead of `height == forks`.

[the remark this answers is not in the document]
Comment (fpelliccioni) [reply src/blockchain/src/populate/populate_base.cpp]: Fixed in PR #192 (merged into master). This branch now includes the fix after rebase — `populate_pooled` correctly compares `stored_height == height` instead of `height == forks`.

[the remark this answers is not in the document]
Comment (fpelliccioni) [reply src/blockchain/src/populate/populate_base.cpp]: Fixed in PR #192 (merged into master). This branch now includes the fix after rebase — `populate_pooled` correctly compares `stored_height == height` instead of `height == forks`.

[the remark this answers is not in the document]
Comment (fpelliccioni) [reply src/blockchain/src/populate/populate_base.cpp]: Fixed in PR #192 (merged into master). This branch now includes the fix after rebase — `populate_pooled` correctly compares `stored_height == height` instead of `height == forks`.

File: src/domain/test/chain/script_vm_auto_4.cpp [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+
+// SKIPPED: 64-bit arithmetic overflow - not implemented
+// TEST_CASE("VM-AUTO #460 [Should Fail: INVALID_NUMBER_RANGE_64_BIT]: 64-bit integer overflow detected", "[vm][auto]") { run_bchn_test({"9223372036854775807", "add1 1 equal", kth::domain::machine::rule_fork::bch_64bit_integers | kth::domain::machine::rule_fork::bch_strictenc | kth::domain::machine::rule_fork::bip16_rule, kth::error::invalid_operand_size, "64-bit integer overflow detected"}); }
+
+// SKIPPED: 64-bit arithmetic overflow - not implemented
+// TEST_CASE("VM-AUTO #461 [Should Fail: INVALID_NUMBER_RANGE_64_BIT]: 9223372036854775807 | add1 1", "[vm][auto]") { run_bchn_test({"9223372036854775807", "add1 1", kth::domain::machine::rule_fork::bch_64bit_integers | kth::domain::machine::rule_fork::bch_strictenc | kth::domain::machine::rule_fork::bip16_rule, kth::error::invalid_operand_size, ""}); }
```
[the remark this answers is not in the document]
Comment (fpelliccioni) [reply src/domain/test/chain/script_vm_auto_4.cpp]: check again

File: src/domain/test/chain/script_vm_auto_14.cpp [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+
+// SKIPPED: MINIMALNUM validation not yet implemented for split/num2bin numeric arguments
+// TEST_CASE("VM-AUTO #1429 [Should Fail: MINIMALNUM]: [0000] [0100] | split 2drop 1", "[vm][auto]") { run_bchn_test({"[0000] [0100]", "split 2drop 1", kth::domain::machine::rule_fork::bch_minimaldata, kth::error::minimal_number, ""}); }
+
+TEST_CASE("VM-AUTO #1430 [Should Pass]: first operand is an unlimited number and exempted from length/minimal encoding rules", "[vm][auto]") { run_bchn_test({"[11111111111111111111111111111111111111110000] [14]", "num2bin drop 1", kth::domain::machine::rule_fork::bch_minimaldata, kth::error::success, "first operand is an unlimited number and exempted from length/minimal encoding rules"}); } // flags: MINIMALDATA, expected: OK
+
+// SKIPPED: MINIMALNUM validation not yet implemented for split/num2bin numeric arguments
+// TEST_CASE("VM-AUTO #1431 [Should Fail: MINIMALNUM]: [11111111111111111111111111111111111111110000] [1400] | num2bin drop 1", "[vm][auto]") { run_bchn_test({"[11111111111111111111111111111111111111110000] [1400]", "num2bin drop 1", kth::domain::machine::rule_fork::bch_minimaldata, kth::error::minimal_number, ""}); }
```
[the remark this answers is not in the document]
Comment (fpelliccioni) [reply src/domain/test/chain/script_vm_auto_14.cpp]: check again

File: src/domain/test/chain/script_vm_auto_15.cpp [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+
+// SKIPPED: CLEANSTACK not implemented yet - requires individual fork activation instead of cumulative upgrades
+// TEST_CASE("VM-AUTO #1569 [Should Fail: CLEANSTACK]: P2SH with unnecessary input", "[vm][auto]") { run_bchn_test({"11 [3044022065066ad31463bd8ebb9cb7ef621e9c96fa3411a3ed393a3e41120a7ea688babf02206869ac4e007f853c1a15bca2670eabeb9a54cac1bddf5c4f73c03f8c17cbffa701] [410479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8ac]", "hash160 [31edc23bdafda4639e669f89ad6b2318dd79d032] equal", kth::domain::machine::rule_fork::bch_cleanstack | kth::domain::machine::rule_fork::bip16_rule, kth::error::cleanstack, "P2SH with unnecessary input"}); }
```
[the remark this answers is not in the document]
Comment (fpelliccioni) [reply src/domain/test/chain/script_vm_auto_15.cpp]: check again

File: src/domain/test/chain/script_vm_auto_8.cpp [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+
+TEST_CASE("VM-AUTO #897 [Should Pass]: REVERSEBYTES, 4 bytes", "[vm][auto]") { run_bchn_test({"[deadbeef]", "reversebytes [efbeadde] equal", kth::domain::machine::rule_fork::bip16_rule, kth::error::success, "REVERSEBYTES, 4 bytes"}); } // flags: P2SH, expected: OK
+
+// SKIPPED: Complex Schnorr signature + NULLFAIL case - BCHN expects success but signature appears invalid, needs investigation of NULLFAIL handling
+// TEST_CASE("VM-AUTO #898 [Should Pass]: REVERSEBYTES, reversed pubkey", "[vm][auto]") { run_bchn_test({"[1052a02f76cf6c3bc43e4671af4c6065f6d918c79f58ad7f90a49c4e2603a4728fc6d81d49087dd97bc43ae184da55dc5195553526ce076df092be27a398b68f41]", "[a8f459fbe84e0568ff03454d438183b0464e0c18ff73658ec0d2619b20ff29e502] reversebytes checksig", kth::domain::machine::rule_fork::bch_nullfail | kth::domain::machine::rule_fork::bch_sighash_forkid | kth::domain::machine::rule_fork::bch_strictenc, kth::error::success, "REVERSEBYTES, reversed pubkey"}); }
```
[the remark this answers is not in the document]
Comment (fpelliccioni) [reply src/domain/test/chain/script_vm_auto_8.cpp]: check again

File: src/domain/test/chain/script_vm_auto_4.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (fpelliccioni) [reply src/domain/test/chain/script_vm_auto_4.cpp]: you was right, but now tests `VM-AUTO #460` and `VM-AUTO #461` are enabled

File: src/domain/test/chain/script_vm_auto_17.cpp [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+
+TEST_CASE("VM-AUTO #1709 [Should Fail: NEGATIVE_LOCKTIME]: CSV automatically fails if stack top is negative", "[vm][auto]") { run_bchn_test({"-1", "checksequenceverify", kth::domain::machine::rule_fork::bip112_rule, kth::error::negative_locktime, "CSV automatically fails if stack top is negative"}); } // flags: CHECKSEQUENCEVERIFY, expected: NEGATIVE_LOCKTIME
+
+TEST_CASE("VM-AUTO #1710 [Should Fail: MINIMALNUM]: CSV fails if stack top is not minimally encoded", "[vm][auto]") { run_bchn_test({"[00]", "checksequenceverify", kth::domain::machine::rule_fork::bch_minimaldata | kth::domain::machine::rule_fork::bip112_rule, kth::error::minimal_number, "CSV fails if stack top is not minimally encoded"}); } // flags: CHECKSEQUENCEVERIFY,MINIMALDATA, expected: MINIMALNUM
+
+TEST_CASE("VM-AUTO #1711 [Should Fail: UNSATISFIED_LOCKTIME]: CSV fails if stack top bit 1 << 31 is set and the tx version < 2", "[vm][auto]") { run_bchn_test({"0", "checksequenceverify", kth::domain::machine::rule_fork::bip112_rule, kth::error::unsatisfied_locktime, "CSV fails if stack top bit 1 << 31 is set and the tx version < 2"}); } // flags: CHECKSEQUENCEVERIFY, expected: UNSATISFIED_LOCKTIME
```
[the remark this answers is not in the document]
Comment (fpelliccioni) [reply src/domain/test/chain/script_vm_auto_17.cpp]: This comment comes from the original BCHN `script_tests.json` data, not from our generator. We verified that BCHN's `BuildSpendingTransaction` also hardcodes `nVersion = 1`, so the test fails due to `tx.nVersion < 2` regardless of the stack value — the "bit 1<<31 is set" part of the comment is misleading in BCHN's own test vectors. Since we faithfully reproduce their test suite, we keep the original description as-is.

File: src/domain/test/chain/script_vm_auto_8.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (fpelliccioni) [reply src/domain/test/chain/script_vm_auto_8.cpp]: Test #898 is already active and running — it is not commented out. The bot appears to be referring to a previous revision of the code where it may have been skipped.

File: src/domain/test/chain/script_vm_auto_14.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (fpelliccioni) [reply src/domain/test/chain/script_vm_auto_14.cpp]: These tests (#1429, #1431) are no longer skipped — they are active and running. The bot is referring to a previous revision of the code.

File: test/scripts/generate_script_tests.py [hunk 14]
```diff
@@ -457,22 +542,44 @@ def replace_pushdata1(match):
...
+        data_hex = match.group(2)
+        return f"[2.{data_hex}]"
+    script = re.sub(inline_pd2, replace_inline_pd2, script)
+
+    # PUSHDATA4 as 0x4eNNNNNNNN 0xDATA
+    inline_pd4 = r'0x4e([0-9a-fA-F]{8})\s+0x([0-9a-fA-F]+)'
+    def replace_inline_pd4(match):
+        data_hex = match.group(2)
+        return f"[4.{data_hex}]"
+    script = re.sub(inline_pd4, replace_inline_pd4, script)
```
[the remark this answers is not in the document]
Comment (fpelliccioni) [reply test/scripts/generate_script_tests.py]: Fixed in the generator: `PUSHDATA` replacements now validate that the declared size matches the actual data length before normalizing. If they don't match, the original (malformed) token sequence is preserved. No change in generated output because current BCHN test vectors don't contain size-mismatched PUSHDATA encodings, but the guard is in place for future vectors.

File: test/scripts/generate_script_tests.py [hunk 15]
```diff
@@ -738,83 +858,299 @@ def process_test_entry(entry: List, line_number: int, knuth_overrides: dict = No
...
+                            direction = "OK->FAIL" if orig_is_ok else "FAIL->OK"
+                            print(f"ERROR: Override changes test semantics ({direction})! "
+                                  f"sig=\"{script_sig}\", pub=\"{script_pub_key}\", "
+                                  f"original={original_expected_error}, override={override_error}",
+                                  file=sys.stderr)
+                            sys.exit(1)
+
                         error_cpp = override_error_cpp
                         all_errors.extend(override_error_errors)
                         override_comment_parts.append(f"ERROR: {override_error}")
```
[the remark this answers is not in the document]
Comment (fpelliccioni) [reply test/scripts/generate_script_tests.py:935]: Fixed — `expected_error` is now updated alongside error_cpp when an override is applied. Please re-review.

File: test/scripts/generate_script_tests.py [hunk 16]
```diff
@@ -738,83 +858,299 @@ def process_test_entry(entry: List, line_number: int, knuth_overrides: dict = No
...
+        if errors:
+            f.write(f"\nPARSING ERRORS ({len(errors)}):\n")
+            for error in errors:
+                f.write(f"  {error}\n")
+        if skip_tests:
+            f.write(f"\nSKIPPED TESTS ({len(skip_tests)}):\n")
+            for i, skip_test in enumerate(skip_tests, 1):
+                f.write(f"  {i}. sig=\"{skip_test.get('script_sig', '')}\" pub=\"{skip_test.get('script_pub_key', '')}\" reason={skip_test.get('reason', '')}\n")
+
+    return generated_files
```
[the remark this answers is not in the document]
Comment (fpelliccioni) [reply test/scripts/generate_script_tests.py:1181]: Fixed — the generator now removes all existing script_vm_auto_*.cpp files before writing new ones. Please re-review.

File: src/blockchain/test/vmlimits_tests.cpp [hunk 17]
```diff
@@ -0,0 +1,487 @@
...
+    int64_t const c_composite_nonstd = c_oc_n + c_hi_n * 64 + int64_t(c_sc_n) * 26'000;
+    CHECK(c_composite_nonstd == expected_op_cost_nonstd);
+
+    // --- Cross-check: native vs consensus ---
+    CHECK(n_sc == c_sc_s);
+    CHECK(n_hi == c_hi_s);
+    CHECK(n_oc == c_oc_s);
+    CHECK(n_stack == c_stack_s);
+#endif // WITH_CONSENSUS
+}
```
[the remark this answers is not in the document]
Comment (fpelliccioni) [reply src/blockchain/test/vmlimits_tests.cpp:188]: Valid point. Tracked in #193 — will extract shared helpers into a common header in a follow-up.

File: src/blockchain/src/validate/validate_input.cpp [hunk 18]
```diff
@@ -16,88 +16,104 @@ namespace kth::blockchain {
...
 
-    if (script::is_enabled(forks, domain::machine::rule_fork::bch_descartes)) {
+    if (script::is_enabled(active_flags, domain::machine::script_flags::bch_p2sh_32)) {
         flags |= verify_flags_enable_p2sh_32;
+    }
+    if (script::is_enabled(active_flags, domain::machine::script_flags::bch_tokens)) {
         flags |= verify_flags_enable_tokens;
     }
 
-    if (script::is_enabled(forks, domain::machine::rule_fork::bch_galois)) {
```
[the remark this answers is not in the document]
Comment (fpelliccioni) [reply src/blockchain/src/validate/validate_input.cpp:80]: Already fixed in the current code — bch_input_sigchecks → verify_flags_input_sigchecks mapping is present at lines 81-83 of validate_input.cpp. 
Please re-review.

File: src/blockchain/test/vmlimits_tests.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (fpelliccioni) [reply src/blockchain/test/vmlimits_tests.cpp:188]: Valid point. Tracked in #194.
